src/NuGet.Core/NuGet.Commands/SourcesCommands/SourcesRunner.cs

```diff
@@ -0,0 +1,299 @@
+            if (existingSourceWithName != null)
+            {
+                args.LogError(string.Format(CultureInfo.CurrentCulture,
+                                Strings.SourcesCommandUniqueName));
+            }
+            var existingSourceWithSource = args.SourceProvider.GetPackageSourceBySource(args.Source);
+            if (existingSourceWithSource != null)
+            {
+                args.LogError(string.Format(CultureInfo.CurrentCulture,
+                                Strings.SourcesCommandUniqueSource));
+            }
+
+            var newPackageSource = new Configuration.PackageSource(args.Source, args.Name);
+
+            if (!string.IsNullOrEmpty(args.Username))
+            {
+                var credentials = Configuration.PackageSourceCredential.FromUserInput(
+                    args.Name,
+                    args.Username,
+                    args.Password,
+                    args.StorePasswordInClearText,
+                    args.ValidAuthenticationTypes);
+                newPackageSource.Credentials = credentials;
+            }
+
+            args.SourceProvider.AddPackageSource(newPackageSource);
+            args.LogMinimal(string.Format(CultureInfo.CurrentCulture,
+                    Strings.SourcesCommandSourceAddedSuccessfully, args.Name));
+        }
+
+        private static void UpdatePackageSource(SourcesArgs args)
+        {
+            if (string.IsNullOrEmpty(args.Name))
+            {
+                args.LogError(string.Format(CultureInfo.CurrentCulture,
+                                Strings.SourcesCommandNameRequired));
+            }
+
+            var existingSource = args.SourceProvider.GetPackageSourceByName(args.Name);
+            if (existingSource == null)
+            {
+                args.LogError(string.Format(CultureInfo.CurrentCulture,
+                                Strings.SourcesCommandNoMatchingSourcesFound, args.Name));
+            }
+
+            if (!string.IsNullOrEmpty(args.Source) && !existingSource.Source.Equals(args.Source, StringComparison.OrdinalIgnoreCase))
+            {
+                if (!PathValidator.IsValidSource(args.Source))
+                {
+                    args.LogError(string.Format(CultureInfo.CurrentCulture,
+                                Strings.SourcesCommandInvalidSource));
+                }
+
+                // If the user is updating the source, verify we don't have a duplicate.
+                var duplicateSource = args.SourceProvider.GetPackageSourceBySource(args.Source);
+                if (duplicateSource != null)
+                {
+                    args.LogError(string.Format(CultureInfo.CurrentCulture,
+                                Strings.SourcesCommandUniqueSource));
+                }
+
+                existingSource = new Configuration.PackageSource(args.Source, existingSource.Name);
+            }
+
+            ValidateCredentials(args);
+
+            if (!string.IsNullOrEmpty(args.Username))
```

Comment:
If password is null, we are going to log the error. This logic will then throw an exception for the same thing inside of `FromUserInput`. Do we want that, or do we want a nice graceful error and then HALT ?

Reply:
boilerplate reply: the code in SourcesRunner.cs was moved from SourcesCommand.cs - I have tried to make the minimal changes to it. That said, I'll review each of these comments. thx.

Reply:
yes, i got carried away with my Logging infrastructure when I was doing this and lost the exceptions. Nice catch. Moving exception type to nuget.commands.dll in nuget.commands namespace and using it again here.
fix in next commit.